src/cdk/a11y/key-manager/tree-key-manager.ts

```diff
@@ -0,0 +1,559 @@
+   * the lifetime of the Tree component.
+   *
+   * Intended to be used to focus the first item in the tree.
+   */
+  onInitialFocus(): void;
+
+  /**
+   * Focus the provided item by index.
+   *
+   * Updates the state of the currently active item. Emits to `change` stream if active item
+   * Changes.
+   * @param index The index of the item to focus.
+   * @param options Additional focusing options.
+   */
+  focusItem(index: number, options?: {emitChangeEvent?: boolean}): void;
+  /**
+   * Focus the provided item.
+   *
+   * Updates the state of the currently active item. Emits to `change` stream if active item
+   * Changes.
+   * @param item The item to focus. Equality is determined via the trackBy function.
+   * @param options Additional focusing options.
+   */
+  focusItem(item: T, options?: {emitChangeEvent?: boolean}): void;
+  focusItem(itemOrIndex: number | T, options?: {emitChangeEvent?: boolean}): void;
+}
+
+export type TreeKeyManagerFactory<T extends TreeKeyManagerItem> = (
+  items: Observable<T[]> | QueryList<T> | T[],
+  options: TreeKeyManagerOptions<T>,
+) => TreeKeyManagerStrategy<T>;
+
+/**
+ * This class manages keyboard events for trees. If you pass it a QueryList or other list of tree
+ * items, it will set the active item, focus, handle expansion and typeahead correctly when
+ * keyboard events occur.
+ */
+export class TreeKeyManager<T extends TreeKeyManagerItem> implements TreeKeyManagerStrategy<T> {
+  private _activeItemIndex = -1;
+  private _activeItem: T | null = null;
+  private _activationFollowsFocus = false;
+  private _horizontal: 'ltr' | 'rtl' = 'ltr';
+  private readonly _letterKeyStream = new Subject<string>();
+  private _typeaheadSubscription = Subscription.EMPTY;
+
+  /**
+   * Predicate function that can be used to check whether an item should be skipped
+   * by the key manager. By default, disabled items are skipped.
+   */
+  private _skipPredicateFn = (item: T) => this._isItemDisabled(item);
+
+  /** Function to determine equivalent items. */
+  private _trackByFn: (item: T) => unknown = (item: T) => item;
+
+  /** Buffer for the letters that the user has pressed when the typeahead option is turned on. */
+  private _pressedLetters: string[] = [];
+
+  private _items: T[] = [];
+
+  constructor(
+    items: Observable<T[]> | QueryList<T> | T[],
+    {
+      skipPredicate,
+      trackBy,
+      horizontalOrientation,
+      activationFollowsFocus,
+      typeAheadDebounceInterval,
+    }: TreeKeyManagerOptions<T>,
+  ) {
+    // We allow for the items to be an array or Observable because, in some cases, the consumer may
+    // not have access to a QueryList of the items they want to manage (e.g. when the
+    // items aren't being collected via `ViewChildren` or `ContentChildren`).
+    if (items instanceof QueryList) {
+      this._items = items.toArray();
+      items.changes.subscribe((newItems: QueryList<T>) => {
```

Comment:
Should this be unsubscribed somewhere, like in a `detach` function?

Reply:
as long as the QueryList gets destroyed, it should automatically unsubscribe

for the `Observable` below, we may want to either handle the subscription manually or document that the provided `Observable` should `complete` when the containing component is destroyed.